planar/triangulate/delaunaytriangulationbuilder.go

```diff
@@ -0,0 +1,214 @@
+	"github.com/go-spatial/geom/cmp"
+	"github.com/go-spatial/geom/planar/triangulate/quadedge"
+)
+
+/*
+A utility class which creates Delaunay Triangulations
+from collections of points and extract the resulting
+triangulation edges or triangles as geometries.
+
+Author Martin Davis
+Ported to Go by Jason R. Surratt
+*/
+type DelaunayTriangulationBuilder struct {
+	siteCoords []quadedge.Vertex
+	tolerance  float64
+	subdiv     *quadedge.QuadEdgeSubdivision
+}
+
+type PointByXY []quadedge.Vertex
+
+func (xy PointByXY) Less(i, j int) bool { return cmp.XYLessPoint(xy[i], xy[j]) }
+func (xy PointByXY) Swap(i, j int)      { xy[i], xy[j] = xy[j], xy[i] }
+func (xy PointByXY) Len() int           { return len(xy) }
+
+/*
+extractUniqueCoordinates extracts the unique points from the given Geometry.
+
+geom - the geometry to extract from
+Returns a List of the unique Coordinates
+*/
+func (dtb *DelaunayTriangulationBuilder) extractUniqueCoordinates(g geom.Geometry) ([]quadedge.Vertex, error) {
+	if g == nil {
+		return []quadedge.Vertex{}, nil
+	}
+
+	coords, err := geom.GetCoordinates(g)
+	if err != nil {
+		return nil, err
+	}
+
+	vertices := make([]quadedge.Vertex, len(coords))
+	for i := range coords {
+		vertices[i] = quadedge.Vertex{coords[i][0], coords[i][1]}
+	}
+
+	return dtb.unique(vertices), nil
+}
+
+func (dtb *DelaunayTriangulationBuilder) unique(points []quadedge.Vertex) []quadedge.Vertex {
+	sort.Sort(PointByXY(points))
+
+	// we can use a slice trick to avoid copying the array again. Maybe better
+	// than two index variables...
+	uniqued := points[:0]
+	for i := 0; i < len(points); i++ {
+		if i == 0 || cmp.PointEqual(points[i], points[i-1]) == false {
+			uniqued = append(uniqued, points[i])
+		}
+	}
+
+	return uniqued
+}
+
+/**
+ * Converts all {@link Coordinate}s in a collection to {@link Vertex}es.
+ * @param coords the coordinates to convert
+ * @return a List of Vertex objects
+public static List toVertices(Collection coords)
+{
+	List verts = new ArrayList();
+	for (Iterator i = coords.iterator(); i.hasNext(); ) {
+		Coordinate coord = (Coordinate) i.next();
+		verts.add(new Vertex(coord));
+	}
+	return verts;
+}
+*/
+
+/**
+ * Computes the {@link Envelope} of a collection of {@link Coordinate}s.
+ *
+ * @param coords a List of Coordinates
+ * @return the envelope of the set of coordinates
+public static Envelope envelope(Collection coords)
+{
+	Envelope env = new Envelope();
+	for (Iterator i = coords.iterator(); i.hasNext(); ) {
+		Coordinate coord = (Coordinate) i.next();
+		env.expandToInclude(coord);
+	}
+	return env;
+}
+*/
+
+/*
+Sets the sites (vertices) which will be triangulated.
+All vertices of the given geometry will be used as sites.
+
+geom - the geometry from which the sites will be extracted.
+*/
+func (dtb *DelaunayTriangulationBuilder) SetSites(g geom.Geometry) error {
+	// remove any duplicate points (they will cause the triangulation to fail)
+	c, err := dtb.extractUniqueCoordinates(g)
```

Comment:
We should either check to see if `dtb` is nil, or we should document that a nil dtb will panic if it is not nil.

Reply:
OK. In nearly all cases I think a panic is appropriate as a nil doesn't have any meaning in this context. I will document accordingly.